core/src/main/java/org/infinispan/xsite/irac/IracBackOffHandler.java

```diff
@@ -0,0 +1,73 @@
+package org.infinispan.xsite.irac;
+
+import java.util.Collection;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.function.Function;
+
+import org.infinispan.commons.util.Immutables;
+import org.infinispan.commons.util.concurrent.CompletableFutures;
+import org.infinispan.util.ExponentialBackOff;
+import org.infinispan.xsite.XSiteBackup;
+
+/**
+ * A global back-off handler for IRAC.
+ * </p>
+ * The back-off time is applied globally for sites put in back-off mode, but each site is independent of the other.
+ * That is, sites `A` and `B` can be reached and receive requests while site `C` is on back-off. Once site `B` is
+ * unreachable and is put to back-off mode, both `B` and `C` receive the same delay. The delay only resets when
+ * the back-off is reset.
+ * </p>
+ * Once a site is put to back-off mode, a {@link CompletableFuture} is scheduled to asynchronously deactivate the
+ * mode using the given {@link Executor}. Once back-off is disabled, it will synchronously activate all sites and
+ * reset the delay time.
+ *
+ * @since 15.0
+ */
+public class IracBackOffHandler implements Function<Void, CompletableFuture<Void>> {
+
+   private final Set<XSiteBackup> backups;
+   private final Executor executor;
+   private final ExponentialBackOff backOff;
+   private volatile CompletableFuture<Void> runnable;
+   private volatile boolean putForBackOff;
+
+   public IracBackOffHandler(Collection<IracXSiteBackup> backups, ExponentialBackOff backOff, Executor executor) {
+      this.backups = Immutables.immutableSetWrap(Set.copyOf(backups));
+      this.backOff = backOff;
+      this.executor = executor;
+      this.runnable = CompletableFutures.completedNull();
+   }
+
+   public CompletableFuture<Void> enableBackOff() {
+      putForBackOff = true;
+      return handleBackOff();
+   }
+
+   public void disableBackOff() {
+      putForBackOff = false;
+      handleBackOff();
+   }
+
+   private CompletableFuture<Void> handleBackOff() {
+      if (putForBackOff) {
```

Comment:
I'm not sure why you need the boolean. `enableBackOff` can inline the code inside this `if` branch and `disableBackOff` invoke the `else` branch.